### PR TITLE
KFSPTS-30797 PDP Extract Unique Filename and Logging

### PR DESCRIPTION
**Pull request by ajd299**

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuExtractPaymentServiceImpl.java [hunk 1]
```diff
@@ -98,12 +98,14 @@ public void extractAchPayments() {
...
     
         String filename = getOutputFile(achFilePrefix, processDate);
-        LOG.debug("MOD: extractAchPayments() filename = " + filename);
 
         /** 
         * MOD: This is the only section in the method that is changed.  This mod calls a new method that bundles 
         * ACHs into single disbursements if the flag to do so is turned on.
+        * Modified the SimpleDateFormat to have milliseconds to prevent files being overwritten
         */
+        sdf = new SimpleDateFormat("yyyy-MM-dd-HH-mm-ss-SSS", Locale.US);
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/pdp/batch/service/impl/CuExtractPaymentServiceImpl.java:107]: I don't think this particular date format needs updating, because this line of code should be unreachable when ISO 20022 formatting is enabled.  (The `if` block above with the `shouldUseIso20022Format()` check should force the method to return early.) Also, the downstream code might be ignoring or replacing this format anyway.

I believe the correct way to modify the filename for the legacy Check/ACH files is to change the date formatter in the `getOutputFile()` method. That method is in base code but can be overridden by this subclass. (Note that we're still generating legacy KFS-proprietary Check files in addition to ISO Check files, so such an override will be needed until we turn off the KFS-proprietary Check file handling in a future sprint.)

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java [hunk 2]
```diff
@@ -375,6 +375,7 @@ public void extractChecks(
                     message.setCstmrCdtTrfInitn(customerCreditTransferInitiation);
 
                     final String filename = determineFilename(directoryName, extractTypeContext);
+                    LOG.info("extractChecks writing file " + filename);
```
Comment (cah292) [inline src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:378]: There's an equivalent spot in the `extractAchs()` method where another logging statement should be inserted. Alternatively, you could move the custom Check/ACH filename logging to the `writeMessageToFile()` method.

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java [hunk 3]
```diff
@@ -1815,7 +1816,7 @@ private String determineFilename(
         }
 
         final Date disbursementDate = extractTypeContext.getDisbursementDate();
-        final SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US);
+        final SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmssSSS", Locale.US);
```
Comment (cah292) [inline src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:1819]: Adjusting this format to include milliseconds won't be enough on its own to guarantee filename uniqueness. Based on the `extractAchs()` and `extractChecks()` methods, the date that's used in the filename is being defined BEFORE iterating over the file creation loops, meaning this filename-building method will likely be reusing the same source date for a given set of Check/ACH files.

To get the filename uniqueness to work as expected, you may need to make one or more of the following changes:

* Redefine the source date for each iteration of the Check/ACH `for` loops. (Not sure whether this would interfere with the data being sent to the bank, though.)
* Modify the `determineFilename()` method so that it also examines the filesystem to see if the given file already exists. If it does, then make adjustments to the filename accordingly, such as by tweaking the millisecond portion or adding/adjusting a custom suffix.

**Review by cah292 — CHANGES_REQUESTED** [review]
I noticed a few things that will need further updates, especially with regards to the filename uniqueness feature.

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java [hunk 4]
```diff
@@ -1815,16 +1815,29 @@ private String determineFilename(
...
 
         final Date disbursementDate = extractTypeContext.getDisbursementDate();
-        final SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US);
+        final String filename = buildUniqueFilename(directoryName, finalFormattedCheckFilePrefix, disbursementDate, null);
+
+        LOG.debug("determineFilename(...) - : filename={}", filename);
+        return filename;
+    }
+
+    private String buildUniqueFilename(String directoryName, String finalFormattedCheckFilePrefix, Date disbursementDate, Integer uniquenessSuffix) {
```
Comment (cah292) [inline src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:1824]: Since this file is an overlay of base code, please add appropriate comments to point out this customization to it.

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java [hunk 5]
```diff
@@ -1815,16 +1815,29 @@ private String determineFilename(
...
                 directoryName
                 + File.separator
                 + finalFormattedCheckFilePrefix
                 + "_"
                 + sdf.format(disbursementDate)
+                + (uniquenessSuffix != null ? "-" + uniquenessSuffix : "")
                 + ".xml";
 
-        LOG.debug("determineFilename(...) - : filename={}", filename);
+        if ((new File(filename).exists())) {
```
Comment (cah292) [inline src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:1836]: There appears to be one extra unnecessary set of parentheses on this line.

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java [hunk 6]
```diff
@@ -1815,16 +1815,29 @@ private String determineFilename(
...
+            LOG.info("buildUniqueFilename filename={} already exists, adding uniqueness suffix", filename);
+            return buildUniqueFilename(directoryName, finalFormattedCheckFilePrefix, disbursementDate, ++uniquenessSuffix);
```
Comment (cah292) [inline src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:1838]: Your initial call to this method sets `uniquenessSuffix` to null. Will that cause the "++" part on this line to encounter a null pointer?

You may need to make some more adjustments to keep this null-safe.

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java [hunk 7]
```diff
@@ -1895,6 +1908,7 @@ private static void writeMessageToFile(
              * of changing/removing the namespace prefixes, rather than performing String.replaceAll() substitution.
              * Also added setup of a CU-specific EscapeHandler to customize the XML-escaping process.
              */
+            LOG.info("writeMessageToFile writing file " + filename);
```
Comment (cah292) [inline src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:1911]: I would suggest modifying this log statement to use a placeholder for inserting the filename.

**Review by cah292 — CHANGES_REQUESTED** [review]
Not sure of this was ready for review again, but I did notice a few minor follow-up items that may need to be addressed.

**Comment by ajd299** [comment]
Please let me know if I have addressed all of the feedback thus far.

**Comment by cah292** [comment]
retest this please

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java [hunk 8]
```diff
@@ -1815,16 +1815,31 @@ private String determineFilename(
...
                 + "_"
                 + sdf.format(disbursementDate)
+                + (uniquenessSuffix != null ? "-" + uniquenessSuffix : "")
                 + ".xml";
 
-        LOG.debug("determineFilename(...) - : filename={}", filename);
+        if (new File(filename).exists()) {
+            LOG.info("buildUniqueFilename filename={} already exists, adding uniqueness suffix", filename);
+            int suffixCounter = uniquenessSuffix == null ? 0 : ++uniquenessSuffix;
+            return buildUniqueFilename(directoryName, finalFormattedCheckFilePrefix, disbursementDate, suffixCounter);
```
Comment (cah292) [inline src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:1840]: This back-and-forth shifting between primitive and wrapper integers is a little confusing. You may want to consider using just a primitive integer instead, set it 0 or 1 as the initial value, and exclude it from the filename if it's currently at the starter value. It might also be helpful to rename the variable to "fileIndex" or something along those lines.

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java (hunk 7 above)
Comment (cah292) [reply src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:1913]: The placeholder suggestion for this log statement has not been implemented yet.

**Review by cah292 — CHANGES_REQUESTED** [review]
There are a couple more areas that might need revisions, but otherwise the changes look good.

**Comment by ajd299** [comment]
Please see my latest changes

File: src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java [hunk 9]
```diff
@@ -1815,16 +1815,30 @@ private String determineFilename(
...
+    private String buildUniqueFilename(String directoryName, String finalFormattedCheckFilePrefix, Date disbursementDate, int fileIndex) {
+        final SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmssSSS", Locale.US);
+
         final String filename =
                 directoryName
                 + File.separator
                 + finalFormattedCheckFilePrefix
                 + "_"
                 + sdf.format(disbursementDate)
+                + (fileIndex == 0 ? "" : ("-" + fileIndex ))
```
Comment (cah292) [inline src/main/java/org/kuali/kfs/pdp/batch/service/impl/Iso20022FormatExtractor.java:1834]: Unnecessary whitespace prior to the closing parentheses.

**Review by cah292 — CHANGES_REQUESTED** [review]
There's one very minor formatting item that would be good to clean up, but otherwise the changes look good.

**Review by cah292 — APPROVED** [review]
Changes look good, unit tests pass and my concerns have been addressed. Merging.
